arangod/GeneralServer/GeneralServer.cpp

```diff
@@ -71,9 +91,33 @@ void GeneralServer::startListening() {
 }
 
 void GeneralServer::stopListening() {
+  for (auto& task : _listenTasks) {
+    task->stop();
+  }
+  
+  // close connections of all socket tasks so the tasks will
+  // eventually shut themselves down
+  MUTEX_LOCKER(lock, _tasksLock);
+  for (auto& task : _commTasks) {
+    task.second->closeStream();
+  }
+}
+
+void GeneralServer::stopWorking() {
   for (auto& context : _contexts) {
     context.stop();
   }
+  
+  _listenTasks.clear();
+
+  while (true) {
+    MUTEX_LOCKER(lock, _tasksLock);
+    if (_commTasks.empty()) {
+      break;
+    }
+    LOG_DEVEL << "waiting for " << _commTasks.size() << " to shut down";
+    std::this_thread::sleep_for(std::chrono::microseconds(5));
```

Comment:
Does this hold the lock during sleep?

Reply:
Oops, I think I also need to remove the log message here.